torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp

```diff
@@ -2170,6 +2311,618 @@ TEST_F(NVFuserTest, FusionCrossIterationGroupedGridAllreduce4_CUDA) {
+  //   0       1           2        3        4         5        6         7
+
+  tv0->reorder(
+      {{3, 0}, {1, 1}, {4, 2}, {7, 3}, {0, 4}, {2, 5}, {5, 6}, {6, 7}});
+  // [iO, r56{persist}, unswitch, iV{2}, rO{BIDy}, rI{TIDy}, iO{BIDx},
+  // i16{TIDx}]
+
+  tv0->axis(4)->parallelize(ParallelType::BIDy);
+  tv0->axis(5)->parallelize(ParallelType::TIDy);
+  tv0->axis(6)->parallelize(ParallelType::BIDx);
+  tv0->axis(7)->parallelize(ParallelType::TIDx);
+  tv0->axis(2)->parallelize(ParallelType::Unswitch);
+
+  TransformPropagator propagator(tv0);
+  MaxRootDomainInfoSpanningTree(tv0).traverse(&propagator);
+
+  auto tv3_rf = tv3->rFactor({1});
+  scheduler_utils::parallelizeAllLike(tv0, ir_utils::allTvs(&fusion));
+
+  // std::cout<<tv3_rf->toString()<<std::endl;
+  // std::cout<<tv3->toString()<<std::endl;
+
+  tv0_cache->axis(3)->parallelize(ParallelType::Vectorize);
+  tv7->axis(3)->parallelize(ParallelType::Vectorize);
+  tv3->axis(2)->parallelize(ParallelType::Group);
+
+  // Inline input into cache, this is our persistent buffer, so don't go inside
+  // persistent dimension
+  auto tv0_cache_consumers = ir_utils::consumerTvsOf(tv0_cache);
+  for (auto consumer : tv0_cache_consumers) {
+    tv0_cache->computeAt(consumer, 1);
+  }
+
+  // Inline into reduction with best effort, this will get through reductions
+  // without reordering them
+  for (auto consumer : tv0_cache_consumers) {
+    if (!DependencyCheck::isDependencyOf(consumer, tv3_rf)) {
+      continue;
+    }
+    consumer->computeAt(tv3_rf, -1, ComputeAtMode::BestEffort);
+  }
+
+  // Inline rfactor into reduction
+  tv3_rf->computeAt(tv3, 1);
+
+  // Inline reduction into broadcast (if this is inlined then grid reduction
+  // would be inside the persistent buffer for loop)
+  tv3->computeAt(tv4, 1);
+
+  // Inline broadcast and later into outputs
+  tv4->computeAt(tv7, -1, ComputeAtMode::MostInlined);
+
+  // // This forces the grid reduction into the persistent loop which gives
+  // terrible perf tv3->computeAt(tv7, -1, ComputeAtMode::MostInlined);
+
+  // Inline residual path with output
+  tv5->computeAt(tv7, -1, ComputeAtMode::MostInlined);
+
+  fusion.printMath();
+  fusion.printKernel();
+
+  auto options_half = at::TensorOptions().dtype(at::kHalf).device(at::kCUDA, 0);
+  at::manual_seed(0);
+
+  const std::vector<int64_t> input_shape{N, H, W, C};
+  auto t0 = at::randn(input_shape, options_half);
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {t0});
+  auto cg_outputs = fe.runFusion({t0});
+  for (int i = 0; i < 10; ++i) {
+    clearL2Cache();
+    cg_outputs = fe.runFusion({t0});
+  }
+
+  auto t1 = t0.to(at::kFloat);
+  auto t2 = t1.sum({0, 1, 2});
+  auto t3 = t2.unsqueeze(0).unsqueeze(0).unsqueeze(0);
+  auto t4 = t1 - t3;
+
+  testValidate(&fusion, cg_outputs, {t0}, {t4}, __LINE__, __FILE__, "");
+}
+
+// Simple persistent kernel on outer normalization for schedule testing.
+TEST_F(NVFuserTest, FusionGridPersistOuterNormLike2_CUDA) {
```

Comment:
@naoyam This is roughly structured how I would imagine it'd be structured through the normalization scheduler. Seems to get decent performance on my machine. This should roughly be the reference we use. I think we need to be careful on how we're doing grouping because the max factor is 8 including multiple reduction expressions right? So when we're doing backward batch norm and have to group the 2 reductions the max vectorization factor would be 4 in fp16?

I think we should next try batch norm backwards for the horizontal fused grid reduction (in fp16). Then we should start working on getting these changes into the normalization scheduler.

Reply:
On my machine (Titan RTX), it resulted in spilling. The effective bandwidth was around 380 GB/s. Not spilling without unswitch, and the bandwidth got bumped up to 440 GB/s

It seems to happen pretty frequently to have much more increased register usages with unswitch. We may need different magic-zero-like hacks to prevent it, but for now, unswitch doesn't seem beneficial for persistent outer reductions.